test/tls-provider.c

```diff
@@ -748,89 +1288,2265 @@ static const OSSL_DISPATCH xor_keymgmt_functions[] = {
+
+    ctx->mdctx = NULL;
+    ctx->md = md;
+    OPENSSL_strlcpy(ctx->mdname, mdname, sizeof(ctx->mdname));
+    return 1;
+}
+
+static int xor_sig_signverify_init(void *vpxor_sigctx, void *vxorsig,
+                                   int operation)
+{
+    PROV_XORSIG_CTX *pxor_sigctx = (PROV_XORSIG_CTX *)vpxor_sigctx;
+
+    if ( pxor_sigctx == NULL || vxorsig == NULL)
+        return 0;
+    xor_freekey(pxor_sigctx->sig);
+    if (!xor_key_up_ref(vxorsig)) return 0;
+    pxor_sigctx->sig = vxorsig;
+    pxor_sigctx->operation = operation;
+    if ( (operation==EVP_PKEY_OP_SIGN && !pxor_sigctx->sig) ||
+         (operation==EVP_PKEY_OP_VERIFY && !pxor_sigctx->sig)) {
+        ERR_raise(ERR_LIB_USER, XORPROV_R_INVALID_KEY);
+        return 0;
+    }
+    return 1;
+}
+
+static int xor_sig_sign_init(void *vpxor_sigctx, void *vxorsig,
+                             const OSSL_PARAM params[])
+{
+    return xor_sig_signverify_init(vpxor_sigctx, vxorsig, EVP_PKEY_OP_SIGN);
+}
+
+static int xor_sig_verify_init(void *vpxor_sigctx, void *vxorsig,
+                               const OSSL_PARAM params[])
+{
+    return xor_sig_signverify_init(vpxor_sigctx, vxorsig, EVP_PKEY_OP_VERIFY);
+}
+
+static int xor_sig_sign(void *vpxor_sigctx, unsigned char *sig, size_t *siglen,
+                    size_t sigsize, const unsigned char *tbs, size_t tbslen)
+{
+    PROV_XORSIG_CTX *pxor_sigctx = (PROV_XORSIG_CTX *)vpxor_sigctx;
+    XORKEY* xorkey = pxor_sigctx->sig;
+
+    size_t max_sig_len = EVP_MAX_MD_SIZE;
+    unsigned int xor_sig_len = 0;
+    int rv = 0;
+
+    if (!xorkey || !xorkey->hasprivkey) {
+      ERR_raise(ERR_LIB_USER, XORPROV_R_NO_PRIVATE_KEY);
+      return rv;
+    }
+
+    if (sig == NULL) {
+      *siglen = max_sig_len;
+      return 1;
+    }
+    if (*siglen < max_sig_len) {
+        ERR_raise(ERR_LIB_USER, XORPROV_R_BUFFER_LENGTH_WRONG);
+        return rv;
+    }
+
+    /* create HMAC using XORKEY as key and hash as data:
+     * No real crypto, just for test, don't do this at home!
+     */
+    if (!HMAC(EVP_sha1(), xorkey->privkey, XOR_KEY_SIZE,
+              tbs, tbslen, &sig[0], &xor_sig_len)) {
+        ERR_raise(ERR_LIB_USER, XORPROV_R_SIGNING_FAILED);
+        goto endsign;
+    }
+
+    *siglen = xor_sig_len;
+    rv = 1; /* success */
+
+ endsign:
+    return rv;
+}
+
+static int xor_sig_verify(void *vpxor_sigctx,
+		          const unsigned char *sig, size_t siglen,
+                          const unsigned char *tbs, size_t tbslen)
+{
+    PROV_XORSIG_CTX *pxor_sigctx = (PROV_XORSIG_CTX *)vpxor_sigctx;
+    XORKEY* xorkey = pxor_sigctx->sig;
+    unsigned char resignature[EVP_MAX_MD_SIZE];
+    unsigned int resiglen;
+    int i;
+
+    if (xorkey == NULL || sig == NULL || tbs == NULL) {
+        ERR_raise(ERR_LIB_USER, XORPROV_R_WRONG_PARAMETERS);
+        return 0;
+    }
+
+    /* This is no real verify: just re-sign and compare:
+     * Don't do this at home! Not fit for real use!
+     */
```

Comment:
...

Reply:
fixed in https://github.com/openssl/openssl/pull/19312/commits/0330203fc25c8b2d08c052ef8070ab6f25d17209